### PR TITLE
Add retry to database connection

### LINKED ISSUE #208: Prevent service from crashing on start if database is not connected
- Currently on master, if the service is started and no database connection can be made, the service crashes
- This is fine, but makes it hard to debug ops issues, as on EKS the service crashes and enters a loop of restarting/crashing
- Per @uberscott 

> ..if the db has to be bounced or something it will cause a cascading failure on the k8s system as the containers are restarted (which will probably stabailize but it could make services unavailable for minutes vs seconds)

- So we should introduce some retry logic when attempting to make an initial database connection

Additional context
- Maybe we retry database connections with some type of exponential backoff
  - Thorough logs so that we know what's happening
  - Set some maximum number of attempts -- then just crash the service after that threshold is met

### PR DESCRIPTION
**Pull request by deekerno**
## Description
Adds retry to database connection attempt so that the indexer service does not immediately crash on start if a connection can't be made. Closes #208.

### CODE REVIEW

**Comment by ra0x3** [comment]
@deekerno Doesn't seem to be working for me? Testing steps below below. Let me know if I've missed something 🤔 


```bash
# stop postgres
➜ brew services start postgresql

# try to start the indexer
➜ git checkout origin/deekerno/add-database-retry-208 && RUST_LOG=info cargo run --bin fuel-indexer
Note: switching to 'origin/deekerno/add-database-retry-208'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by switching back to a branch.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -c with the switch command. Example:

  git switch -c <new-branch-name>

Or undo this operation with:

  git switch -

Turn off this advice by setting config variable advice.detachedHead to false

HEAD is now at d7e5a02 Add retry to database connection
    Finished dev [unoptimized + debuginfo] target(s) in 0.52s
     Running `target/debug/fuel-indexer`
Sep 22 15:22:12.079  INFO fuel_indexer: Configuration: IndexerConfig { fuel_node: FuelNodeConfig { host: "127.0.0.1", port: "4000" }, graphql_api: GraphQLConfig { host: "127.0.0.1", port: "29987", run_migrations: None }, database: PostgresConfig { user: "postgres", password: "XXXX", host: "127.0.0.1", port: "5432", database: "postgres" } }
thread 'main' panicked at 'Failed to establish postgres connection.: Io(Os { code: 61, kind: ConnectionRefused, message: "Connection refused" })', /Users/rashad/Development/repos/fuel-indexer/fuel-indexer-database/postgres/src/lib.rs:36:10
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Review by ra0x3 — CHANGES_REQUESTED** [review]
@deekerno Doesn't seem to be working for me? Testing steps below below. Let me know if I've missed something 🤔 


```bash
# stop postgres
➜ brew services stop postgresql

# try to start the indexer
➜ git checkout origin/deekerno/add-database-retry-208 && RUST_LOG=info cargo run --bin fuel-indexer
Note: switching to 'origin/deekerno/add-database-retry-208'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by switching back to a branch.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -c with the switch command. Example:

  git switch -c <new-branch-name>

Or undo this operation with:

  git switch -

Turn off this advice by setting config variable advice.detachedHead to false

HEAD is now at d7e5a02 Add retry to database connection
    Finished dev [unoptimized + debuginfo] target(s) in 0.52s
     Running `target/debug/fuel-indexer`
Sep 22 15:22:12.079  INFO fuel_indexer: Configuration: IndexerConfig { fuel_node: FuelNodeConfig { host: "127.0.0.1", port: "4000" }, graphql_api: GraphQLConfig { host: "127.0.0.1", port: "29987", run_migrations: None }, database: PostgresConfig { user: "postgres", password: "XXXX", host: "127.0.0.1", port: "5432", database: "postgres" } }
thread 'main' panicked at 'Failed to establish postgres connection.: Io(Os { code: 61, kind: ConnectionRefused, message: "Connection refused" })', /Users/rashad/Development/repos/fuel-indexer/fuel-indexer-database/postgres/src/lib.rs:36:10
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Comment by deekerno** [comment]
> Doesn't seem to be working for me? Testing steps below below. Let me know if I've missed something 🤔

@ra0x3 That's odd because it connects to the database just fine for me.

```
    Finished dev [unoptimized + debuginfo] target(s) in 13.14s
     Running `target/debug/fuel-indexer`
Sep 22 18:37:41.682  INFO fuel_indexer: Configuration: IndexerConfig { fuel_node: FuelNodeConfig { host: "127.0.0.1", port: "4000" }, graphql_api: GraphQLConfig { host: "127.0.0.1", port: "29987", run_migrations: None }, database: PostgresConfig { user: "postgres", password: "XXXX", host: "127.0.0.1", port: "5432", database: "postgres" } }
Sep 22 18:37:41.862  INFO sqlx::query: SELECT current_database(); rows affected: 0, rows returned: 1, elapsed: 5.225ms  
Sep 22 18:37:41.865  INFO sqlx::query: SELECT pg_advisory_lock($1); rows affected: 1, rows returned: 1, elapsed: 2.819ms  
Sep 22 18:37:41.868  INFO sqlx::postgres::notice: relation "_sqlx_migrations" already exists, skipping    
Sep 22 18:37:41.872  INFO sqlx::query: CREATE TABLE IF NOT …; rows affected: 0, rows returned: 0, elapsed: 2.121ms
...

The only difference between our two tests is that I'm running Docker through Postgres. 🤔 
```

**Comment by deekerno** [comment]
```
❯ RUST_LOG=debug cargo run --bin fuel-indexer
    Blocking waiting for file lock on build directory
   Compiling fuel-indexer-postgres v0.1.0 (/Users/deekerno/dev/fuel/indexer/fuel-indexer-database/postgres)
   Compiling fuel-indexer-database v0.1.0 (/Users/deekerno/dev/fuel/indexer/fuel-indexer-database)
   Compiling fuel-indexer-schema v0.0.0 (/Users/deekerno/dev/fuel/indexer/fuel-indexer-schema)
   Compiling fuel-indexer v0.0.0 (/Users/deekerno/dev/fuel/indexer/fuel-indexer)
    Finished dev [unoptimized + debuginfo] target(s) in 11.34s
     Running `target/debug/fuel-indexer`
Sep 23 01:30:08.581  INFO fuel_indexer: Configuration: IndexerConfig { fuel_node: FuelNodeConfig { host: "127.0.0.1", port: "4000" }, graphql_api: GraphQLConfig { host: "127.0.0.1", port: "29987", run_migrations: None }, database: PostgresConfig { user: "postgres", password: "XXXX", host: "127.0.0.1", port: "5432", database: "postgres" } }
Sep 23 01:30:08.719  WARN fuel_indexer_postgres: Could not connect to postgres backend, retrying in 15 seconds...
Sep 23 01:30:23.889  WARN fuel_indexer_postgres: Could not connect to postgres backend, retrying in 30 seconds...
Sep 23 01:30:54.096  WARN fuel_indexer_postgres: Could not connect to postgres backend, retrying in 60 seconds...
Sep 23 01:31:54.288  WARN fuel_indexer_postgres: Could not connect to postgres backend, retrying in 120 seconds...
Sep 23 01:33:54.471  WARN fuel_indexer_postgres: Could not connect to postgres backend, retrying in 240 seconds...
thread 'main' panicked at 'Retry attempts exceeded; could not connect to postgres backend!', /Users/deekerno/dev/fuel/indexer/fuel-indexer-database/postgres/src/lib.rs:54:21
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
👍🏽

File: fuel-indexer-database/sqlite/src/lib.rs [hunk 1]
```diff
@@ -1,7 +1,11 @@
 use fuel_indexer_database_types::*;
 use fuel_indexer_lib::utils::sha256_digest;
 use sqlx::{pool::PoolConnection, types::JsonValue, Connection, Row, Sqlite, SqliteConnection};
-use tracing::info;
+use tokio::time::{sleep, Duration};
+use tracing::{info, warn};
+
+const RETRY_LIMIT: usize = 5;
+const INITIAL_RETRY_DELAY: u64 = 15;
```
Comment (ra0x3) [inline fuel-indexer-database/sqlite/src/lib.rs:8]: I would maybe lower this to something like `2`. If the DB doesn't connect in 30 seconds we have other issues. Not a blocker

**Review by ra0x3 — APPROVED** [review]
Seems fine. I would maybe create an issue and tag it with `code quality` to see if we can cleanup some of the duplicate logic.

**Review by ra0x3 — APPROVED** [review]

**Pull request merged**